### PR TITLE
enh : Memorize Duplicate `intrinsicsElementalFunction` + `intrinsicsArrayFunction`

### LINKED ISSUE #4318: Bug : intrinsics applied to array inside `where` raises error after applying 'array_op' asr_pass
**MRE**
```Fortran
program name
    implicit none
    integer :: a(3)
    integer  :: b(3)
    a = [1,0,-1]
    b = 0
    where(abs(a) == 1) b = 555
    print *, b       ! Expected Result : 555 0 555  
end program name

```
Error

```bash
ASR verify pass error: The variable in ArrayItem must be an array, not a scalar
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/assem/lfortran_test/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/assem/lfortran_test/lfortran/src/bin/lfortran.cpp", line 2481, in ??
    return main_app(argc, argv);
  File "/home/assem/lfortran_test/lfortran/src/bin/lfortran.cpp", line 2442, in main_app(int, char**)
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/assem/lfortran_test/lfortran/src/bin/lfortran.cpp", line 954, in ??
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/assem/lfortran_test/lfortran/src/lfortran/fortran_evaluator.cpp", line 363, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/assem/lfortran_test/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 9967, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, co.po, diagnostics);
  File "/home/assem/lfortran_test/lfortran/src/libasr/pass/pass_manager.h", line 316, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    apply_passes(al, asr, _passes, pass_options, diagnostics);
LCompilersException: Verify failed in the pass: array_op

```

Found in files :
`outer.f90` line 341
`inner.f90` line 331

### PR DESCRIPTION
**Pull request by assem2002**
Fixes #4318 

I noticed that there could be some duplicated node for `intrinsicsElementalFunction` (nodes with same pointer address). They're mainly made by `where` asr_pass, which is great as we can memorize their result and just whenever we encounter them again we can simply get their result without doing all this work.
***

**Why is that related to the issue with intrinsics call inside where ?**

The reason is that after `where` pass did it work it left us with 3 duplicated `intrinsicsElementalFunction`. Whenever you visit a single one of them and you get any of the internal nodes tweaked or replaced, the others get affected as well which results in error in the great helper phase `asr_verify` as one of the node has done its replace phase correctly but the others changed and now they don't fulfill their purpose (whether being inside `arraybound` or `integercompare` ).
***
**Concerns and suggestions**

- I don't know if we should fix the error from `array_op` pass OR to create independent nodes while replacing `where` with `doLoop` in `where` pass.
Notice that :  `intrinsicsElementalFunction` could be a node representing `m_v` for `lbound` + `Ubound` and the replacment for these would be creating 2 independent `__libasr__created__var__X__elemental_func_call_res` and creating do loops for both , with the help of memorizing we won't worry about that.

- If it's a valid idea we can emplace the result in `m_value` of  `intrinsicsElementalFunction`, and whenever we replace `intrinsicsElementalFunction` we would check its value member.

### CODE REVIEW

File: src/libasr/pass/array_op.cpp [hunk 1]
```diff
@@ -75,6 +75,7 @@ class ReplaceArrayOp: public ASR::BaseExprReplacer<ReplaceArrayOp> {
     ASR::dimension_t* op_dims; size_t op_n_dims;
     ASR::expr_t* op_expr;
     std::map<ASR::expr_t*, ASR::expr_t*>& resultvar2value;
+    std::map<void*, ASR::expr_t*> intrinsics2result_var;
```
Comment (certik) [inline src/libasr/pass/array_op.cpp:78]: What type is the first type? If it is an expression, let's use `expr_t` instead of void.

Comment (assem2002) [reply src/libasr/pass/array_op.cpp:78]: The first type could be either `intrinsicsElementalFunction` or `intrinsicsArrayFunction`, so I used `void*`.
You can consider the idea of emplacing the result in `m_value` member, I think they are always `nullptr` then we can make good use of it and get rid of the overhead resulting from using `map`

Comment (certik) [reply src/libasr/pass/array_op.cpp:78]: Both `IntrinsicElementalFunction` and `IntrinsicArrayFunction` are `expr`, so it should be `ASR::expr_t*`.

Comment (assem2002) [reply src/libasr/pass/array_op.cpp]: Yeah that's my bad I should've used `expr_t*`, I'm pushing commit with the changes now.

**Comment by assem2002** [comment]
@certik  The commits now should be good.
The main branch now resolves the issue we're trying to fix with using intrinsics inside `where` due to the fix in this PR https://github.com/lfortran/lfortran/pull/4360. The PR is solving the problem by creating indpenedant `intrinsicElementalFunction` nodes in the `where` pass instead of using the same nodes. The solution is quite valid, but it faces the problem of creating duplicate loops that aren't necessary.


**Example :**
```fortran
program t
    implicit none
    integer ::  a(3)
    where (abs(a) == 1 ) a = 555
end program t
``` 

**How main handles it now :**
```fortran 

! Fortran code after applying the pass: array_op
program name
implicit none
integer(4) :: __1_k
integer(4) :: __1_t
integer(4) :: __1_v
integer(4), dimension(3) :: __libasr__created__var__0__elemental_func_call_res
integer(4), dimension(3) :: __libasr__created__var__1__elemental_func_call_res
integer(4), dimension(3) :: __libasr__created__var__2__elemental_func_call_res
integer(4), dimension(3) :: a
__1_v = lbound(a, 1)
do __1_t = lbound(__libasr__created__var__0__elemental_func_call_res, 1),&
         ubound(__libasr__created__var__0__elemental_func_call_res, 1)
    __libasr__created__var__0__elemental_func_call_res(__1_t) = _lcompilers_abs_i32(a(__1_v))
    __1_v = __1_v + 1
end do
__1_v = lbound(a, 1)
do __1_t = lbound(__libasr__created__var__1__elemental_func_call_res, 1),&
         ubound(__libasr__created__var__1__elemental_func_call_res, 1)
    __libasr__created__var__1__elemental_func_call_res(__1_t) = _lcompilers_abs_i32(a(__1_v))
    __1_v = __1_v + 1
end do
do __1_k = lbound(__libasr__created__var__0__elemental_func_call_res, 1),&
         ubound(__libasr__created__var__1__elemental_func_call_res, 1), 1
    __1_v = lbound(a, 1)
    do __1_t = lbound(__libasr__created__var__2__elemental_func_call_res, 1),&
         ubound(__libasr__created__var__2__elemental_func_call_res, 1)
        __libasr__created__var__2__elemental_func_call_res(__1_t) = _lcompilers_abs_i32(a(__1_v))
        __1_v = __1_v + 1
    end do
    if (__libasr__created__var__2__elemental_func_call_res(__1_k) == 1) then
        a(__1_k) = 555
    end if
end do

contains

integer(4) function _lcompilers_abs_i32(x)
    integer(4), intent(in) :: x
    if (x >= 0) then
        _lcompilers_abs_i32 = x
    else
        _lcompilers_abs_i32 = -x
    end if
end function _lcompilers_abs_i32

end program name

```


**While Using the same nodes result in this output :**

```fortran

! Fortran code after applying the pass: array_op
program name
implicit none
integer(4) :: __1_k
integer(4) :: __1_t
integer(4) :: __1_v
integer(4), dimension(3) :: __libasr__created__var__0__elemental_func_call_res
integer(4), dimension(3) :: a
__1_v = lbound(a, 1)
do __1_t = lbound(__libasr__created__var__0__elemental_func_call_res, 1),&
         ubound(__libasr__created__var__0__elemental_func_call_res, 1)
    __libasr__created__var__0__elemental_func_call_res(__1_t) = _lcompilers_abs_i32(a(__1_v))
    __1_v = __1_v + 1
end do
do __1_k = lbound(__libasr__created__var__0__elemental_func_call_res, 1),&
         ubound(__libasr__created__var__0__elemental_func_call_res, 1), 1
    if (__libasr__created__var__0__elemental_func_call_res(__1_k) == 1) then
        a(__1_k) = 555
    end if
end do

contains

integer(4) function _lcompilers_abs_i32(x)
    integer(4), intent(in) :: x
    if (x >= 0) then
        _lcompilers_abs_i32 = x
    else
        _lcompilers_abs_i32 = -x
    end if
end function _lcompilers_abs_i32

end program name

```

**Comment by assem2002** [comment]
We can change that later maybe when we do major ASR pass refactoring in the code base, if that's the case then we can close this PR for now or open an issue related to this PR. 
If we wish to avoid redundant do loop creation, We should use the same nodes when we're calling `visit_where()` in where pass.

**Comment by certik** [comment]
Can you fix the conflict and let's merge it.

**Comment by assem2002** [comment]
I'll resolve it and revert the changes mentioned here https://github.com/lfortran/lfortran/pull/4334#issuecomment-2186294751.

**Comment by assem2002** [comment]
It's ready for review.

- Reverted creating independent nodes.
- Resolved conflicts with main.

**Review by certik — APPROVED** [review]
I think that this is fine, thanks for fixing it!

**Pull request merged**